RecoMET/METProducers/src/PFMETProducer.cc

```diff
@@ -95,12 +120,50 @@ namespace cms {
 
     //Compute the covariance matrix and fill it
     double sumPtUnclustered = 0;
-    reco::METCovMatrix cov = metSigAlgo_->getCovariance(
-        *inputJets, leptons, candInput, *rho, resPtObj, resPhiObj, resSFObj, event.isRealData(), sumPtUnclustered);
+    reco::METCovMatrix cov;
+    if (applyWeight_)
+      cov = metSigAlgo_->getCovariance(*inputJets,
+                                       leptons,
+                                       candInput,
+                                       *rho,
+                                       resPtObj,
+                                       resPhiObj,
+                                       resSFObj,
+                                       event.isRealData(),
+                                       sumPtUnclustered,
+                                       weights_);
+    else
+      cov = metSigAlgo_->getCovariance(
+          *inputJets, leptons, candInput, *rho, resPtObj, resPhiObj, resSFObj, event.isRealData(), sumPtUnclustered);
 
     return cov;
   }
 
+  void PFMETProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+    desc.add<edm::InputTag>("src", edm::InputTag("particleFlow"));
+    desc.add<bool>("calculateSignificance", false);
+    desc.add<double>("globalThreshold", 0.);
+    desc.add<std::string>("alias", "pfMet");
+    desc.addOptional<edm::InputTag>("srcJets");
+    desc.addOptional<std::vector<edm::InputTag>>("srcLeptons");
+    desc.addOptional<std::string>("srcJetSF");
+    desc.addOptional<std::string>("srcJetResPt");
+    desc.addOptional<std::string>("srcJetResPhi");
+    desc.addOptional<edm::InputTag>("srcRho");
+    edm::ParameterSetDescription params;
+    params.setAllowAnything();
```

Comment:
it seems like the format is rather fixed. Are these always required (just with a swap of data vs MC)?
If so, it seems better to define these without defaults so that user code is forced to use them

Reply:
the params are needed when calculateSignificance=True. What do you suggest to add/change?

Reply:
there is a "switch" option in the PSet validation https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Switch

Reply:
hmm, with a switch, they are required to be present when calculateSignficicance is true, but the plugin doesn't accept them given when calculateSignficance is false.
how do I implement "required if true, optional if false"?

Reply:
I thought that the logic is "required if true; not needed otherwise".
It sounds like the calculateSignficicance is not controlling anything and the parameter is needed anyways.
My earlier comment was about the setAllowAnything. Perhaps you can anyways define the structure (drop setAllowAnything) and leave it optional?

Reply:
I thought that the logic is "required if true; not needed otherwise".
It sounds like the calculateSignficicance is not controlling anything and the parameter is needed anyways.
My earlier comment was about the setAllowAnything. Perhaps you can anyways define the structure (drop setAllowAnything) and leave it optional?

Reply:
if calculateSignficance: srcJets, ..., parameters are required.
if not calculateSignficance: srcJets, ..., parameters are not needed, but there is also no reason to forbit them.
I don't know how to implement this

Reply:
ok, so not use switch and go back to addOptional. understood.

Reply:
params.setAllowAnything(); on the other hand is not used directly in this module, but in another cc file. can look into this later.

Reply:
> params.setAllowAnything(); on the other hand is not used directly in this module, but in another cc file. can look into this another time.

it's a fair point that the outside "plugin" should provide this description (to be inserted here). I'm fine for a later PR, but if updates are coming, please add a "FIXME" here in the comments.
Thank you.